server/main.rs

```diff
@@ -1,22 +1,411 @@
+#![feature(core)]
 #![feature(io)]
+#![feature(net)]
+#![feature(path)]
+#![feature(std_misc)]
 
 extern crate "hematite_server" as hem;
+extern crate uuid;
+extern crate rand;
+extern crate time;
 
+use std::io::prelude::*;
 use std::io;
+// use std::iter::repeat;
+use std::net::{TcpListener, TcpStream};
+use std::old_io::timer::sleep;
+use std::ops::Sub; // Sub for Timespec
+use std::path::Path;
+use std::sync::Arc;
+use std::thread;
+use std::time::Duration;
 
-use hem::packet::Protocol;
+use hem::packet::{self, PacketRead, PacketWrite, Protocol, NextState};
+use hem::proto::ping;
+use hem::proto::properties::Properties;
+use hem::types::{
+    // BlockPos,
+    // Chat,
+    Chunk, ChunkColumn,
+    // EMPTY_SLOT, Slot,
+    Var
+};
+use hem::types::consts::*;
+use hem::util::ReadExactExt;
 
-fn main() {
-    let value = false;
-    let first_value = vec![0];
+use uuid::Uuid;
 
-    let mut w = Vec::new();
-    <bool as Protocol>::proto_encode(&value, &mut w).unwrap();
-    assert_eq!(&w, &first_value);
+const PACKET_NAMES: [&'static str; 26] = [
+    "(c2s) KeepAlive",
+    "(c2s) ChatMessage",
+    "(c2s) UseEntity",
+    "(c2s) Player",
+    "(c2s) PlayerPosition",
+    "(c2s) PlayerLook",
+    "(c2s) PlayerPositionAndLook",
+    "(c2s) PlayerDigging",
+    "(c2s) PlayerBlockPlacement",
+    "(c2s) HeldItemChange",
+    "(c2s) Animation",
+    "(c2s) EntityAction",
+    "(c2s) SteerVehicle",
+    "(c2s) CloseWindow",
+    "(c2s) ClickWindow",
+    "(c2s) ConfirmTransaction",
+    "(c2s) CreativeInventoryAction",
+    "(c2s) EnchantItem",
+    "(c2s) UpdateSign",
+    "(c2s) PlayerAbilities",
+    "(c2s) TabComplete",
+    "(c2s) ClientSettings",
+    "(c2s) ClientStatus",
+    "(c2s) PluginMessage",
+    "(c2s) Spectate",
+    "(c2s) ResourcePackStatus"
+];
 
-    let mut r = io::Cursor::new(w);
-    let value = <bool as Protocol>::proto_decode(&mut r).unwrap();
-    assert_eq!(false, value);
+/// TODO(toqueteos): Move this to its own module. Proposal: src/vanilla/mod.rs
+struct Server {
+    /// World start timestamp
+    start: time::Timespec,
+    /// Number of ticks since world start (1 second = 20 ticks)
+    ticks: u64,
+    /// Dummy player storage, just their username.
+    players: Vec<String>,
+}
+
+impl Server {
+    fn new() -> Server {
+        Server {
+            ticks: 0,
+            players: Vec::new(),
+            start: time::now().to_timespec(),
+        }
+    }
+
+    pub fn world_age(&self) -> i64 {
+        let end = time::now().to_timespec();
+        let elapsed = end.sub(self.start).num_seconds();
+        elapsed * 20
+    }
+
+    pub fn time_of_day(&self) -> i64 {
+        self.world_age() % 24000
```

Comment:
The same could be said about chunk columns, this might not be the way, but it works. There's no notion of world (there's just an enum for dimension) in the codebase, this way we can see what's needed and where.